### ISSUE TITLE
 static Windows build 

### ISSUE THREAD

**Issue by 1268** [ldc-developers/ldc#4499]
Using this program:

~~~d
import stdio = std.stdio;

void main() {
   stdio.writeln("hello world");
}
~~~

and this command:

~~~
ldc2 app.d
~~~

I do not get a static build. Namely, the file `vcruntime140.dll` is required. So I tried this:

~~~
ldc2 -static app.d
~~~

same result. It seems LDC includes the static library:

~~~
    Directory: D:\ldc2\lib\mingw

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----        2022-07-20  1:10 PM          32542 vcruntime140.lib
~~~

so it would appear it should be possible to do what I am trying to do.

**Comment by kinke** [comment]
You need a Visual C++ installation, then the MSVC runtime can be linked statically (and is by default for non-DLLs). See the 2nd-last paragraph in [README.txt](https://github.com/ldc-developers/ldc/blob/master/packaging/README.txt).

**Comment by 1268** [comment]
have you guys considering using another Windows DLL, thats already pre installed? this would solve the issue of having to download a redistributable, or even having to install the full Visual Studio or whatever.

this is whats done by the Go programming language, as well as Rust. this is the one thing keeping me from getting into D development, because this details makes distributing D programs harder than other languages.

**Comment by kinke** [comment]
Oh, it's you again.

**Comment by 1268** [comment]
did you have anything constructive to add? I genuinely like the D language, but this detail is an important one I think. If D has a good reason why it needs the current DLL versus other options, thats fine, but I think the decision should be justified, because it causes an extra burden regarding the distribution of D programs.

**Comment by Imperatorn** [comment]
> have you guys considered using another Windows DLL, thats already pre installed? this would solve the issue of having to download a redistributable, or even having to install the full Visual Studio or whatever.
> 
> this is whats done by the Go programming language, as well as Rust. this is the one thing keeping me from getting into D development, because this detail makes distributing D programs harder than other languages.

I think it's a license issue iirc. Do you know how Go and Rust solves it? 

But the easiest fix is to just have a Visual C++ installation on your system.

**Comment by kassane** [comment]
I carried out some tests using ldc2 + `zig cc` (replacing `clang-cl`) [target **msvc**] and got this result:

https://gist.github.com/kassane/2a181fa4dfe66dbd283f3ad7a941ca2b

**Comment by 3052** [comment]
@kassane I am only seeing the command and its output - can you clarify your conclusion? does it mean with Zig C compiler, you dont need to install Visual C++/Visual Studio? if so that is cool, but I wonder how thats even legally possible

**Comment by Imperatorn** [comment]
Iirc the licensing is tricky. That's why you need the vc redist.

**Comment by 3052** [comment]
> vc redist.

no, that is not valid for this use case. you actually need the development files, not just the runtime files. looking at the comments, it seems he is actually using Xwin to download the Microsoft development files:

    xwin --manifest-version 16 --accept-license splat --output $PWD/.xwin --include-debug-symbols --preserve-ms-arch-notation --include-debug-libs

then Zig to do the actual compiling. thats a really annoying amount of ceremony, but using Xwin is a way better option than trying to install Visual Studio. I looked at Xwin before, but at the time I decided to ignore it, exactly because it didn't include the compiler. so the option is the Gist is not great, but its probably the best option currently, unless someone finds something similar to Xwin that ALSO downloads the Microsoft compiler.

**Comment by kassane** [comment]
> someone finds something similar to Xwin that ALSO downloads the Microsoft compiler.

https://github.com/mstorsjo/msvc-wine?tab=readme-ov-file#use-with-clanglld-in-msvc-mode

**Comment by 3052** [comment]
> https://github.com/mstorsjo/msvc-wine?tab=readme-ov-file#use-with-clanglld-in-msvc-mode

this doesn't seem to be any better than the option presented in the Gist. it just seems to use Clang to compile instead of Zig, which I would guess is actually a larger download. so unless I am misunderstanding, the Gist option while not great, seems to be the best option currently.

**Comment by kassane** [comment]
I agree. Currently, the Zig toolchain has become convenient due to its small size and facilitation.

If LDC chose MinGW, Zig also offers native integration (currently, default target), although it does not have winpthreads included.

However the dependencies below are statically binary linked. This occurs with `zig c++`, but not functional for msvc.

What was presented in the gist would be something nostdlib++.

**C ABI**
src: https://github.com/ziglang/zig/tree/master/lib/libc/mingw

**C++ ABI**/**C++EH ABI**
src: https://github.com/ziglang/zig/tree/master/lib/libcxx
src: https://github.com/ziglang/zig/tree/master/lib/libunwind

I don't know how BetterC supports disabling d_runtime. But Zig also has a custom compiler runtime (ziglang does not have an lib ABI).

LLVM-libcompiler-rt rewritten in zig.
src: https://github.com/ziglang/zig/tree/master/lib/compiler_rt

**Comment by 3052** [comment]
thanks for the info, but that way WAY more info that I want to have to think about. this is exactly why I decide on the Go programming language versus D, because I didn't want to have to mess with this crap. with Go, I just download the one toolchain from the main website, and I can compile fully static executables out of the box. D cannot do that on Windows, never has been able to, and probably never will. so its for that reason that I will stick with Go, or Rust or even Zig.

I would love to use D as its a cool language, but I am not willing to bend over backward just to get a static executable.

**Comment by 3052** [comment]
OK after talking with  Martin Storsjö, it seems `vsdownload.py` does download the Microsoft compiler as well, so that might actually be a better option than the Gist
